### PR TITLE
feat: support lifecycle hooks in bzlmod

### PR DESCRIPTION
**Pull request by kormide**
In support of https://github.com/aspect-build/rules_js/issues/644.

### CODE REVIEW

File: npm/extensions.bzl [hunk 1]
```diff
@@ -53,7 +63,16 @@ def _extension_impl(module_ctx):
...
-            imports = npm_translate_lock_helpers.gen_npm_imports(importers, packages, lock_patched_dependencies, attr.pnpm_lock.package, attr.name, attr, registries, utils.default_registry(), npm_auth)
+
+            lifecycle_hooks, lifecycle_hooks_execution_requirements = macro_helpers.macro_lifecycle_args_to_rule_attrs(
+                attr.lifecycle_hooks,
+                attr.lifecycle_hooks_exclude,
+                attr.run_lifecycle_hooks,
+                attr.lifecycle_hooks_no_sandbox,
+                attr.lifecycle_hooks_execution_requirements,
+            )
+            imports = npm_translate_lock_helpers.gen_npm_imports(importers, packages, lock_patched_dependencies, attr.pnpm_lock.package, attr.name, attr, lifecycle_hooks, lifecycle_hooks_execution_requirements, registries, utils.default_registry(), npm_auth)
```
Comment (gregmagolan) [inline npm/extensions.bzl:74]: eek, the order dependency and the number of args here is scary. perhaps switch to named parameters:
```
...gen_npm_imports(
   importers = importers,
   packages = packages,
   ...
```
?

File: npm/extensions.bzl [hunk 2]
```diff
@@ -53,7 +63,16 @@ def _extension_impl(module_ctx):
             if attr.npmrc:
                 npmrc = parse_npmrc(module_ctx.read(attr.npmrc))
                 (registries, npm_auth) = npm_translate_lock_helpers.get_npm_auth(npmrc, module_ctx.path(attr.npmrc), module_ctx.os.environ)
-            imports = npm_translate_lock_helpers.gen_npm_imports(importers, packages, lock_patched_dependencies, attr.pnpm_lock.package, attr.name, attr, registries, utils.default_registry(), npm_auth)
+
+            lifecycle_hooks, lifecycle_hooks_execution_requirements = macro_helpers.macro_lifecycle_args_to_rule_attrs(
+                attr.lifecycle_hooks,
```
Comment (gregmagolan) [inline npm/extensions.bzl:68]: name parameters here as well?

**Review by gregmagolan — APPROVED** [review]
lgtm; supporting both bzlmod and WORKSPACE for npm_translate_lock has been quite a bit of work hey

File: npm/extensions.bzl [hunk 3]
```diff
@@ -62,9 +93,10 @@ def _extension_impl(module_ctx):
                     custom_postinstall = i.custom_postinstall,
                     deps = i.deps,
                     integrity = i.integrity,
-                    lifecycle_hooks = i.lifecycle_hooks,
+                    lifecycle_hooks = i.run_lifecycle_hooks if i.run_lifecycle_hooks and i.lifecycle_hooks else [],
```
Comment (kormide) [inline npm/extensions.bzl:96]: This was the issue. I had to mimic what was being put into `{maybe_lifecycle_hooks}` which had this conditional.

Comment (gregmagolan) [reply npm/extensions.bzl:96]: 👍

**Pull request merged**